File: 1926-nearest-exit-from-entrance-in-maze/1926-nearest-exit-from-entrance-in-maze.py
class Solution:
    def nearestExit(self, maze: List[List[str]], entrance: List[int]) -> int:
        m, n = len(maze), len(maze[0])
        
        start_row, start_col = entrance[0], entrance[1]
        queue = [(start_row, start_col, 0)]
        maze[start_row][start_col] = '+'
        
        total_step = -1
        directions = [(1,0), (-1,0), (0,1), (0,-1)]
        while (queue):
            row, col, step = queue.pop(0)
            is_in_border = row == 0 or col == 0 or row == m-1 or col == n-1
            is_in_entrance = row == entrance[0] and col == entrance[1]

            if (is_in_border and not is_in_entrance):
                total_step = step
                break

            for direction in directions:
                new_row = row + direction[0]
                new_col = col + direction[1]

                if (new_row >= m or new_col >= n or new_row < 0 or new_col < 0):
                    continue

                if (maze[new_row][new_col] == '.'):
                    queue.append((new_row,new_col, step+1))
                    maze[new_row][new_col] = '+' # visited

        return total_step
        
        

File: 1926-nearest-exit-from-entrance-in-maze/1926-nearest-exit-from-entrance-in-maze.java
class Solution {
    public int nearestExit(char[][] maze, int[] entrance) {
        int m = maze.length; 
        int n = maze[0].length; 
        Queue<int[]> queue = new LinkedList<>();
        int[] startCoor = {entrance[0], entrance[1], 0};
        queue.add(startCoor);

        int[][] directions = {
            {-1, 0}, // UP
            {1, 0},  // DOWN
            {0, -1}, // LEFT
            {0, 1}   // RIGHT
        };

        while (!queue.isEmpty()){
            int[] coordinate = queue.remove();
            int x = coordinate[0];
            int y = coordinate[1];
            int step = coordinate[2];

            System.out.println(Arrays.toString(coordinate));
            
            if (maze[x][y] == '-'){
                continue; // visited
            }

            maze[x][y] = '-';

            if (!(x == entrance[0] && y == entrance[1]) && (x == 0 || y == 0 || x == m-1 || y == n-1)){
                return step;
            }

            for (int[] dir : directions){
                int[] nextCoor = {dir[0] + x, dir[1] + y, step+1};
                if (nextCoor[0] < 0 || nextCoor[0] >= m || nextCoor[1] < 0 || nextCoor[1] >= n)
                    continue;

                if (maze[nextCoor[0]][nextCoor[1]] == '.'){
                    queue.add(nextCoor);
                }
            }
        }


        return -1;
    }

}